modules/geochemistry/src/utils/EquilibriumGeochemicalSystem.C

```diff
@@ -0,0 +1,1310 @@
+void EquilibriumGeochemicalSystem::buildTemperatureDependentQuantities(Real /*temperature*/)
+{
+  // When temperature least-squares stuff has been done, this function will be altered
+  // at 25degC currently
+  for (unsigned eqm_j = 0; eqm_j < _num_eqm; ++eqm_j)
+    _eqm_log10K[eqm_j] = _mgd.eqm_log10K(eqm_j, 1);
+  for (unsigned red = 0; red < _num_redox; ++red)
+    _redox_log10K[red] = _mgd.redox_log10K(red, 1);
+}
+
+void
+EquilibriumGeochemicalSystem::buildAlgebraicInfo(std::vector<bool> & in_algebraic_system,
+                                                 unsigned & num_basis_in_algebraic_system,
+                                                 unsigned & num_in_algebraic_system,
+                                                 std::vector<unsigned> & algebraic_index,
+                                                 std::vector<unsigned> & basis_index) const
+{
+  // build in_algebraic_system
+  for (const auto & name_index : _mgd.basis_species_index)
+  {
+    const std::string name = name_index.first;
+    const unsigned basis_ind = name_index.second;
+    const ConstraintMeaningEnum meaning = _constraint_meaning[basis_ind];
+    if (name == "H2O")
+      in_algebraic_system[basis_ind] = (meaning == ConstraintMeaningEnum::MOLES_BULK_WATER);
+    else if (_mgd.basis_species_gas[basis_ind])
+      in_algebraic_system[basis_ind] = false;
+    else if (_mgd.basis_species_mineral[basis_ind])
+      in_algebraic_system[basis_ind] = false;
+    else
+      in_algebraic_system[basis_ind] = (meaning == ConstraintMeaningEnum::MOLES_BULK_SPECIES);
+  }
+
+  // build algebraic_index and basis_index
+  num_basis_in_algebraic_system = 0;
+  algebraic_index.resize(_num_basis, 0);
+  basis_index.resize(_num_basis, 0);
+  for (unsigned basis_ind = 0; basis_ind < _num_basis; ++basis_ind)
+    if (in_algebraic_system[basis_ind])
+    {
+      algebraic_index[basis_ind] = _num_basis_in_algebraic_system;
+      basis_index[_num_basis_in_algebraic_system] = basis_ind;
+      num_basis_in_algebraic_system += 1;
+    }
+
+  num_in_algebraic_system = num_basis_in_algebraic_system + _num_surface_pot;
+}
+
+unsigned
+EquilibriumGeochemicalSystem::getNumInAlgebraicSystem() const
+{
+  return _num_in_algebraic_system;
+}
+
+unsigned
+EquilibriumGeochemicalSystem::getNumBasisInAlgebraicSystem() const
+{
+  return _num_basis_in_algebraic_system;
+}
+
+unsigned
+EquilibriumGeochemicalSystem::getNumSurfacePotentials() const
+{
+  return _num_surface_pot;
+}
+
+const std::vector<bool> &
+EquilibriumGeochemicalSystem::getInAlgebraicSystem() const
+{
+  return _in_algebraic_system;
+}
+
+const std::vector<unsigned> &
+EquilibriumGeochemicalSystem::getBasisIndexOfAlgebraicSystem() const
+{
+  return _basis_index;
+}
+
+const std::vector<unsigned> &
+EquilibriumGeochemicalSystem::getAlgebraicIndexOfBasisSystem() const
+{
+  return _algebraic_index;
+}
+
+std::vector<Real>
+EquilibriumGeochemicalSystem::getAlgebraicVariableValues() const
+{
+  std::vector<Real> var;
```

Comment:
Just a comment and not a change request: you know the size of these vectors ahead of time! You should size stuff like this at init wherever you can.

Reply:
Even better is to switch to std::array for fixed size "vectors". It can make a big difference as they are stack allocated, not heap allocated so you get better memory locality.

Reply:
In this case, i don't the size of this vector, as it can change during a simulation.  There are other cases that are fixed at instantiation, however, and i can change those if you like.

Reply:
Hmmm, i didn't understand about arrays.  I see their sizes have to be known at compile time, which is not true in this case, boohoo.  So i won't be changing any vectors to arrays